server.c

```diff
@@ -1292,6 +1295,96 @@ client_close(struct client *c)
 	client_close_ev(c->fd, 0, c);
 }
 
+static ssize_t
+read_cb(struct tls *ctx, void *buf, size_t buflen, void *cb_arg)
+{
+	struct client *c = cb_arg;
+
+	if (!c->should_buffer) {
+		// no buffer to cache into, read into libtls buffer
+		ssize_t ret = read(c->fd, buf, buflen);
+		if (-1 == ret && errno == EWOULDBLOCK)
+			ret = TLS_WANT_POLLIN;
+		
+		return ret;
+	}
+
+	if (c->buf.has_tail) {
+		// we have leftover data from a previous call to read_cb
+		size_t left = BUFLAYER_MAX - c->buf.read_pos;
+		size_t copy_len = MINIMUM(buflen, left);
+		memcpy(buf, c->buf.data + c->buf.read_pos, copy_len);
+
+		c->buf.read_pos += copy_len;
+
+		if (left == copy_len) {
+			// memset(buflayer, 0, BUFLAYER_MAX);
+			c->should_buffer = 0;
+			c->buf.has_tail = 0;
+		}
+		
+		return copy_len;
+	}
+
+	// buffer layer exists, we expect proxy protocol
+	errno = 0;
+	ssize_t n_read = read(
+		c->fd, 
+		c->buf.data + c->buf.len, 
+		BUFLAYER_MAX - c->buf.len
+	);
+	if (-1 == n_read && errno == EWOULDBLOCK)
+		return TLS_WANT_POLLIN;
+
+	c->buf.len += n_read;
+
+	struct proxy_protocol_v1 pp1 = {0};
+	size_t consumed = 0;
+	
+	int parse_status = proxy_proto_v1_parse(&pp1, c->buf.data, c->buf.len, &consumed);
+	if (PROXY_PROTO_PARSE_SUCCESS != parse_status) {
+		close(c->fd);
+		return 0;
```

Comment:
missed one thing: `close(c->fd)` here is a bit dangerous because later in `client_close` we'll close it again.  file descriptor numbers can be reused, so we might end up closing another client socket!

So, here we should IMHO log something (e.g. `log_warnx("invalid proxy protocol header")` or something), then return -1.  `tls_read()` will then bubble the error up to libevent which will call `client_error` where we'll finally properly close it.